GPII-1224: Implement the "Hybrid Flow Manager"

simonbates opened the pull request:
https://issues.gpii.net/browse/GPII-1224

Please note that the code is not yet ready for merging into the main project repository.

This pull request has been created for the purpose of early review.

amb26 commented on gpii/node_modules/flowManager/src/FlowManagerUtilities.js:
```diff
@@ -184,6 +185,55 @@
+
+    // TODO: Restore the ability of the ontology manager to straightforwardly form single transforms and remove this hard-coded path
+
+    gpii.request.flowManager.privacy.privacyToSettingsTransformRules = fluid.require("${universal}/testData/ontologies/privacy-flat.json");
```
Did we run into a problem based on this line with respect to running the web-based tests for the lifecycleManager?

simonbates replied:
I haven't encountered any issues with regards to the lifecycleManager web-based tests.

amb26 commented:
Hi @simonbates - I gave the branch a look over and it is looking very good. I think the main work needed to get it merged in is just to improve the factoring with respect to the existing code - that is the material duplicated from the userLogin request and the base FlowManager. Thanks for all the work on this so far!

kaspermarkus commented:
Hi @simonbates - know you're quite pressed on the security work, but when you have some time available, could you look at this pull request? It would be great to get in relatively soon

simonbates commented:
I've updated the branch to implement the new logonChange API and I've added support for anonymous tokens.

As part of the logonChange work, I was able to remove the duplication in the userLogin handler.

amb26 commented on gpii/node_modules/flowManager/test/js/PrivacyFilterTests.js:
```diff
@@ -0,0 +1,215 @@
+                            {
+                                "type": "http://registry.gpii.net/conditions/inRange",
+                                "min": 400,
+                                "inputPath": "http://registry\\.gpii\\.net/common/environment/visual\\.luminance"
```
All instances of this term in trunk should have been changed to "illuminance" rather than "visual\\.luminance" - I think there are a couple of solutions in testData which are still lagging - we should fix these test fixtures so they aren't a source of the old term

simonbates replied:
I've created a JIRA issue for this condition updating: https://issues.gpii.net/browse/GPII-1326

colinbdclark commented:
:tada: Congratulations on this really huge and helpful contribution, @simonbates! :+1:

kaspermarkus commented:
Yay!!! Thanks @simonbates and @amb26 for the work that went into this!
